src/wasm-delegations-fields.def

```diff
@@ -665,8 +693,8 @@ switch (DELEGATE_ID) {
   }
   case Expression::Id::ArrayNewSegId: {
     DELEGATE_START(ArrayNewSeg);
-    DELEGATE_FIELD_INT(ArrayNewSeg, op);
-    DELEGATE_FIELD_NAME(ArrayNewSeg, segment);
+    DELEGATE_FIELD_NAME_KIND(ArrayNewSeg, dataSegment, ModuleItemKind::DataSegment);
+    DELEGATE_FIELD_NAME_KIND(ArrayNewSeg, elemSegment, ModuleItemKind::ElementSegment);
```

Comment:
What if we kept `op` and `segment`, but then used the ternary operator on `op` to determine what `ModuleItemKind` to pass in for `segment`? I'd really like to avoid having two different mutually exclusive names in the class.

Reply:
We can't quite do that here. Note that we don't have a `curr` object at all, as this is all quite abstract, and in fact we may be operating on more than one object at a time (as equality comparisons do). That is, we don't have a way to query a specific instance at this level.

Can you explain why you don't like having mutually exclusive names? If it's size then we do remove the `op` field, which balances things out. And as mentioned above we do have precedent for this. It seems ok to me?

Reply:
Too bad :( It's much nicer when we can use the C++ type system to make it impossible to represent meaningless values rather than having to remember and check a bunch of extra constraints. IMO the cleanest solution here is to split the classes and factor out shared code where possible. I would rather have ~100 extra lines of code than have these mutually exclusive name fields 😅

Reply:
Fair enough, it is something the reader needs to keep in mind. I'll work on a patch.